ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -343,6 +347,71 @@ private void onAccountChangeMessage(String topic, String message) {
+      getValidContainers(accountToContainerMap, statsSnapshot, null);
+      Set<Container> deleteInProgressContainerSet = getContainersByStatus(Container.ContainerStatus.DELETE_IN_PROGRESS);
+      for (Container container : deleteInProgressContainerSet) {
+        String containerIdToString = "C[" + container.getId() + "]";
+        String accountIdToString = "A[" + container.getParentAccountId() + "]";
+        if (accountToContainerMap.containsKey(accountIdToString) && accountToContainerMap.get(accountIdToString).contains(containerIdToString)) {
+          logger.info("Container {} has not been deleted yet", container);
+        } else {
+          invalidContainerCandidateSet.add(container);
+        }
+      }
+    }
+    return invalidContainerCandidateSet;
+  }
+
+  /**
+   * Selects {@link Container}s to be marked as INACTIVE and marked in zookeeper.
+   */
+  public void selectInvalidContainersAndMarkInZK(StatsSnapshot statsSnapshot) {
+    Set<Container> invalidContainerCandidateSet = selectInvalidContainerCandidates(statsSnapshot);
+    markContainerZkNodesInactive(invalidContainerCandidateSet);
+  }
+
+  /**
+   * Mark the given {@link Container}s status to INACTIVE in zookeeper.
+   * @param invalidContainerSet DELETE_IN_PROGRESS {@link Container} set which has been deleted successfully during compaction.
+   */
+  private void markContainerZkNodesInactive(Set<Container> invalidContainerSet) {
+    // TODO: mark the given containers status to INACTIVE in zookeeper.
+  }
+
+  /**
+   * Gets valid data size {@link Container}s. The qualified {@link Container}s' raw valid data size should be larger than zero.
+   * @param accountToContainerMap it holds a mapping of {@link Account}s to {@link Container}s which raw valid data size larger than zero.
+   * @param statsSnapshot the {@link StatsSnapshot} generated from cluster wide aggregation.
+   * @param keyName the key of subMap for each level of {@link StatsSnapshot}.
+   */
+  private void getValidContainers(Map<String, Set<String>> accountToContainerMap, StatsSnapshot statsSnapshot,
+      String keyName) {
+    if (statsSnapshot.getSubMap() == null || statsSnapshot.getSubMap().isEmpty()) {
+      return;
+    } else {
+      for (Map.Entry<String, StatsSnapshot> entry : statsSnapshot.getSubMap().entrySet()) {
+        if (entry.getKey().startsWith("C") && entry.getValue().getValue() > 0) {
+          Preconditions.checkNotNull(keyName,
+              "keyName should not be null since every container will have it's corresponding accountId");
+          accountToContainerMap.getOrDefault(keyName, new HashSet<>()).add(entry.getKey());
+        } else if (entry.getKey().startsWith("A")) {
+          accountToContainerMap.putIfAbsent(entry.getKey(), new HashSet<>());
+          getValidContainers(accountToContainerMap, entry.getValue(), entry.getKey());
+        } else {
+          getValidContainers(accountToContainerMap, entry.getValue(), null);
```

Comment:
I'm thinking that if for the future we probably will have multiple level of subMap under container level. For example like separate the statsReport by month, so with this it will quickly jump out of recursion. If you think it's unnecessary, please let me know. I can remove it.

Reply:
Let's remove it for now. If we really need this, we can add later.

Reply:
Removed.